lighthouse-core/gather/driver.js

```diff
@@ -1332,10 +1334,20 @@ class Driver {
 
   /**
    * @param {LH.Config.Settings} settings
+   * @param {Partial<EmulationParams['metrics']>} [deviceMetricsOverrides] Override device size for taking screenshots.
    * @return {Promise<void>}
    */
-  async beginEmulation(settings) {
-    await emulation.emulate(this, settings);
+  async beginEmulation(settings, deviceMetricsOverrides = {}) {
+    if (settings.emulatedFormFactor && settings.emulatedFormFactor !== 'none') {
```

Comment:
does this mean our fullpage screenshot breaks when `emulatedFormFactor === 'none'`?

I think I'd prefer to have the fullpage screenshot go through protocol directly than `beginEmulation`, maybe we need to write our own equivalent of `getDeviceMetrics` and just reset them to *whatever* they were before we did the screenshot?

Reply:
good point, let's rethink this.

the issue is indeed with how to put emulation back to how it was.

the protocol doesn't allow for partial device metric overrides, so we must supply all the values. `width` and `deviceScaleFactor` depends on settings, height is set to content size. We just want to vary height but we must set all the others too.

using the protocol directly for this is easy. and must be done unconditionally, no matter the value of `emulatedFormFactor` or `internalDisableDeviceScreenEmulation`.

reverting the supertall emulation amounts to calling `beginEmulation(settings)`. Or, if emulation is disabled from a setting, just `Emulation.clearDeviceMetricsOverride`.

that seem right?

Reply:
> that seem right?

Right for everything except the puppeteer cases we're now pushing folks to use in conjunction with lighthouse to solve the custom cases :/

Is there any way to piece together the current device metrics values? I'm assuming something along the lines of an evaluateAsync `window.devicePixelRatio` plus some combination of `innerWidth/innerHeight` and touch event support?

I do wish there was a simple `Emulation.getDeviceMetrics`...

If it's too much of a pain, I'm ok with the clearDeviceMetricsOverride strategy for now but it'd be nice to move in the direction of not assuming we own everything as we try to better support custom flows and such later in the year.

Reply:
> Is there any way to piece together the current device metrics values? I'm assuming something along the lines of an evaluateAsync window.devicePixelRatio plus some combination of innerWidth/innerHeight and touch event support?

Gave that a shot. I think `mobile` and screen orientation will be brittle.

What if we had an option callback in LH runner API to set emulation, so that we don't have to guess when we want to reset things?

Reply:
we really need some good example cases / tests to nail the puppeteer workflow. I'd like to get this to a state where it may be buggy w/ that workflow, and iterate on it later. wdyt?

Reply:
ya that's fine 👍